### PR TITLE
MQTT Revamp

### PR DESCRIPTION
**Pull request by anwarulislam**
Closes hoppscotch/internal-issues#14

### Description
Revamping the UI and subscription to topic improvement in MQTT

### Checks

- [x] My pull request adheres to the code style of this project
- [x] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [x] All the tests have passed

### CODE REVIEW

File: packages/hoppscotch-app/components/realtime/Mqtt.vue [hunk 1]
```diff
@@ -175,25 +415,31 @@ const subscriptionState = useReadonlyStream(
...
-const canPublish = computed(
-  () =>
-    pubTopic.value !== "" &&
-    message.value !== "" &&
-    connectionState.value === "CONNECTED"
-)
-const canSubscribe = computed(
-  () => subTopic.value !== "" && connectionState.value === "CONNECTED"
-)
+const subscriptionModal = ref(false)
```
Comment (AndrewBastin) [inline packages/hoppscotch-app/components/realtime/Mqtt.vue:436]: More descriptive variable names.

```suggestion
const subscriptionModalShown = ref(false)
```

File: packages/hoppscotch-app/components/realtime/Mqtt.vue [hunk 2]
```diff
@@ -32,101 +32,331 @@
...
-          id="publish"
-          name="get"
-          :disabled="!canPublish"
-          :label="$t('mqtt.publish')"
-          @click.native="publish"
-        />
+    </template>
+    <template #secondary>
+      <div class="h-full flex flex-col hide-scrollbar !overflow-auto">
+        <div
```
Comment (AndrewBastin) [inline packages/hoppscotch-app/components/realtime/Mqtt.vue:186]: These tabs stuff should be in a separate component as well.
If the MQTT component is to deal with the rendering the whole of the MQTT UI, Tab logic should not be its concern!!

Single Response Principle please!!!!!

File: packages/hoppscotch-app/components/realtime/Mqtt.vue [hunk 3]
```diff
@@ -345,6 +608,116 @@ const getI18nError = (error: MQTTError): string => {
...
+    changeTab(topic)
+  }
+
+  tabs.value.push({
+    id: topic,
+    name: topic,
+    color,
+    removable: true,
+  })
+  changeTab(topic)
```
Comment (AndrewBastin) [inline packages/hoppscotch-app/components/realtime/Mqtt.vue:660]: This is going to change the tab to the topic id if one of the tabs match. Okay, that sorta makes sense I think, but then there is no returns going on here, the push will happen even if something matched.

**Review by AndrewBastin — CHANGES_REQUESTED** [review]
The Mqtt component should be split apart to respective bits.
These Tabs should become a new independent component itself.

And personally the current UI gives me literally 1 log entry space on the Log component at a time, which is horrible.

File: packages/hoppscotch-app/components/realtime/Subscription.vue [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+  },
+})
+
+const emit = defineEmits<{
+  (e: "hide-modal"): void
+  (e: "submit", body: MQTTTopic): void
+}>()
+
+const QoSValues = [2, 1, 0] as const
+const QoS = ref<2 | 1 | 0>(2)
```
Comment (AndrewBastin) [inline packages/hoppscotch-app/components/realtime/Subscription.vue:121]: Compile time constants should be UPPER_SNAKECASE.

Also, it is better to bind the types of these together.

```suggestion
const QoS_VALUES = [2, 1, 0] as const
const QoS = ref<typeof QoS_VALUES[number]>(2)
```

File: packages/hoppscotch-app/components/realtime/ConnectionConfig.vue [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+const t = useI18n()
+const QoSOptions = ref<any>()
+const QoSValues = [2, 1, 0] as const
+
+const emit = defineEmits<{
+  (e: "change", body: MQTTConnectionConfig): void
+}>()
+
+// config
```
Comment (AndrewBastin) [inline packages/hoppscotch-app/components/realtime/ConnectionConfig.vue:125]: Unhelpful comment

File: packages/hoppscotch-app/components/realtime/ConnectionConfig.vue [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+</template>
+
+<script setup lang="ts">
+import { ref, watch } from "@nuxtjs/composition-api"
+import { useI18n } from "~/helpers/utils/composables"
+import { MQTTConnectionConfig } from "~/helpers/realtime/MQTTConnection"
+
+const t = useI18n()
+const QoSOptions = ref<any>()
+const QoSValues = [2, 1, 0] as const
```
Comment (AndrewBastin) [inline packages/hoppscotch-app/components/realtime/ConnectionConfig.vue:119]: With the repeated usage of QoSValues in some places, I think it is better to move it to the `MQTTConnection` file.

**Review by AndrewBastin — CHANGES_REQUESTED** [review]
Just some minor corrections :)

**Comment by anwarulislam** [comment]
@liyasthomas do you have any feedback regarding MQTT UI?

**Comment by balub** [comment]
@anwarulislam please refer to the notion doc for improvements/issues found with this feature.

**Review by balub — CHANGES_REQUESTED** [review]
@anwarulislam fix the issues with the tabs and other bugs pointed out in the doc.

File: packages/hoppscotch-app/src/components/realtime/ConnectionConfig.vue [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+            :placeholder="t('mqtt.lw_message')"
+          />
+        </div>
+        <div class="flex items-center justify-between px-4">
+          <div class="flex items-center">
+            <label class="font-semibold text-secondaryLight">
+              {{ t("mqtt.lw_qos") }}
+            </label>
+            <tippy
+              ref="QoSOptions"
```
Comment (AndrewBastin) [inline packages/hoppscotch-app/src/components/realtime/ConnectionConfig.vue:64]: This template ref is not used and can be removed.

File: packages/hoppscotch-app/src/components/realtime/Subscription.vue [hunk 8]
```diff
@@ -0,0 +1,141 @@
+<template>
+  <SmartModal v-if="show" dialog :title="t('mqtt.new')" @close="hideModal">
+    <template #body>
+      <div class="flex justify-between mb-4">
+        <div class="flex items-center">
+          <label class="font-semibold text-secondaryLight">
+            {{ t("mqtt.qos") }}
+          </label>
+          <tippy
+            ref="QoSOptions"
```
Comment (AndrewBastin) [inline packages/hoppscotch-app/src/components/realtime/Subscription.vue:10]: Unused ref.

**Review by AndrewBastin — CHANGES_REQUESTED** [review]
Remove unused refs.

Also one helpful refactor that can be done. For the template refs within the files affected, it would be cool if you changed the declaration from `ref<any>(null)` to `ref<TippyComponent>()` (you can import `TippyComponent` from "vue-tippy")

**Comment by balub** [comment]
Screenshot 2022-11-21 125314
@anwarulislam can we make it so that the default background color (light or dark mode) is cannot be used as an option for selection when creating a new subscription .

**Review by balub — COMMENTED** [review]
Screenshot 2022-11-21 125314

@anwarulislam can we make it so that the default background color (light or dark mode) cannot be used as an option for selection when creating a new subscription .

**Comment by anwarulislam** [comment]
@balub, since we are using a color picker to pick the color when subscribing to a topic the user has full control over it. In either case, we can define some constant color and show it to the user to select from. What's your suggestion?

**Comment by liyasthomas** [comment]
I can take care of this. @balub, please approve the PR if everything else looks good to you. So that I can review and append my changes to this branch.

**Review by liyasthomas — APPROVED** [review]
LGTM

**Pull request merged**